### ISSUE TITLE
Configuring with --disable-mpi-fortran still installs mpifort(1)

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#8682]
@dalcinl noticed that if you configure with `--disable-mpi-fortran`, we still install `mpifort` (and sym-linked friends, such as `mpif90`).  This seems incorrect, especially since they include `-lmpi_mpifh`, and `libmpi_mpifh` does not exist (i.e., is not installed).

I observed this on the v4.1 branch, but it's probably in v4.0, master, and v5.0 as well.

@ggouaillardet Can you think of any reason that we would want to install `mpifort` (etc.) when there are no Fortran bindings?  I can't think of any -- I'd just like to check to make sure that there's no dark, subtle reason we're doing this...

**Comment by ggouaillardet** [comment]
@jsquyres that sounds like a bug.

There is a good reason for installing C++ wrappers even when C++ bindings are not built, but I cannot remember a good rationale for doing it with Fortran.

**Comment by dalcinl** [comment]
Well, if MPI had an ABI, we could be using MPI via ISO_C_BINDINGS.

On a more everyday situation, imagine that you have some app written in both C and Fortran. The C code makes MPI calls, but the Fortran code is all sequential, no MPI. At some point you have to link your object files. Some folks will likely prefer to do the linking via Fortran. So it makes sense to use `mpifort`, even if no Fortran MPI is involved. Of course, `mpifort` should link in just the plain "C" MPI library, because the Fortran library is not available.

So, in short, perhaps it is a good idea to still install `mpifort`. 

However, take a look here:
```shell
$ mpifort -show
gfortran -I<prefix>/include -Wl,-rpath -Wl,<prefix>/lib -L<prefix>/lib -lmpi_mpifh -lmpi
```
the `-lmpi_mpifh` is obviously wrong, this compiler wrapper line is broken.

**Comment by jsquyres** [comment]
That may well be the reason we're still installing `mpifort` and friends (to link an overall C+Fortran MPI application).

I don't think I have a strong opinion here; I *believe* that going either way will not be a big fix (i.e., either removing fortran libraries from the Fortran wrapper compiler link lines, or not installing `mpifort` at all).

**Comment by bosilca** [comment]
I like @dalcinl idea because it is a very normal scenario for many HPC applications, and today's solution (aka. finding the correct include/link options to make it work) is a major hassle.

**Comment by jsquyres** [comment]
@bosilca Thanks -- I was going to raise this issue for larger visibility, but since at least one other person chimed in, I'm happy to accept that solution (i.e., let's remove the MPI Fortran libs and make `mpifort` still be safe to use).

FWIW: this has been broken since at least 4.0.0 (I didn't test earlier than that).  So this hasn't worked in a long time.

**Comment by hjelmn** [comment]
I vote for not installing it at all if --disable-mpi-fortran. I run a fortran-free system and ompi-tests makes some assumptions if it sees mpifort at all.

**Comment by dalcinl** [comment]
@hjelmn Wouldn't it be better to update ompi-tests assumptions? Is that really hard?

@jsquyres Alternatively, perhaps you need an extra configure flag `--disable-fortran` that will skip absolutely everything Fortran-related (bindings, wrappers, the stuff ompi_info prints, etc)? This way you accommodate for @hjelmn user case. And now that I'm at it, because asking is free and you do the work, why not `--disable-cxx`?

**Comment by jsquyres** [comment]
> I vote for not installing it at all if --disable-mpi-fortran. I run a fortran-free system and ompi-tests makes some assumptions if it sees mpifort at all.

If that were true, then you wouldn't have been able to build/run ompi-tests on your Fortran-free system for years... 😉

**Comment by hjelmn** [comment]
@jsquyres I have a local copy with all fortran configury hacked out.

**Comment by jsquyres** [comment]
FWIW, I just built `ompi` with `--disable-mpi-fortran` and the ompi-tests IBM suite built ok for me.  Is there something else in `ompi-tests` that doesn't work?

**Comment by hjelmn** [comment]
@jsquyres Do you have a fortran compiler installed? I don't.

**Comment by dalcinl** [comment]
I just tried to run configure in Fedora 33 container (using Fedora's toolbox tool):

1. I assume `./autogen.pl` was already run outside the container, so the configure script is already available in the source tree.
2. I started by installing `make`, `flex`, and `gcc` packages (the `gcc` package brings in just the C compiler)
3. I ran `./configure --disable-man-pages`. It did not work, configure cannot find a C++ compiler.
4. Then I installed the `gcc-c++` package.
5. I ran `./configure --disable-man-pages` again. I works, no complains about a missing Fortran compiler.

As expected, `ompi_info | grep Fort` shows  `no|none` for all entries:

`mpifort` is still installed, however:
```
$ mpifort --show
--------------------------------------------------------------------------
No underlying compiler was specified in the wrapper compiler data file
(e.g., mpicc-wrapper-data.txt)
--------------------------------------------------------------------------
```
and the command exit with error, which I kind of like
```
$ echo $?
1
```

@jsquyres Perhaps you may consider improving point 3. If no C++ compiler is available, configure should either continue, or the user should be told about using a `--disable-cxx` configure option. And you may still considering adding a `--disable-fortran` flag to remove Fortran support in full.

@hjelmn From point 5 above, looks like current master can be built just find without Fortran. Perhaps your hacks are just for ompi-tests?

**Comment by jsquyres** [comment]
@hjelmn Yes, that's why I was asking you what breaks.  😜 

I think there's multiple issues here:

1. If there's no Fortran compiler found, we shouldn't install `mpifort` and friends.
2. If there's a Fortran compiler found but we're not building the Fortran bindings (e.g., `--disable-mpi-fortran`), we should install `mpifort` and friends, but not link against the MPI libraries.
3. If there's a Fortran compiler found and we _are_ building the Fortran bindings, then install the full/complete `mpifort` and friends.

**Comment by jsquyres** [comment]
> I ran ./configure --disable-man-pages. It did not work, configure cannot find a C++ compiler.

Yeah, we should fix that, too.  There's no need for a C++ compiler any more.

**Comment by dalcinl** [comment]
> 2\. install `mpifort` and friends, but have them not link against the MPI libraries.

Small correction: ... not link against the MPI **Fortran** libraries. You still want `mpifort` to link the **C** MPI libraries, such that it can be used as an MPI-aware linker for mixed C+Fortran applications.

**Comment by jsquyres** [comment]
@dalcinl You are correct -- that was a typo in my above comment.

**Comment by jsquyres** [comment]
#8693 is a PR to fix the C++ compiler requirement (which was [not-so-]surprisingly complicated and time consuming 😕  -- that took up my available time today...).  A few elements of that PR will be useful in fixing the Fortran / `mpifort` issue.

**Comment by jsquyres** [comment]
#8716 is the PR to fix the `mpifort` issues (and one minor mistake missed in #8693).

**Comment by jsquyres** [comment]
@dalcinl We have fixed this in master and v5.0.x. Do you have strong feelings about the fix getting back-ported to v4.1.x and v4.0.x?

**Comment by dalcinl** [comment]
> Do you have strong feelings about the fix getting back-ported to v4.1.x and v4.0.x?

Not at all, master and v5.0.x are enough for me.

**Comment by jsquyres** [comment]
Cool; closing this, then.
